### PR TITLE
keyd: add keyd service and test

### PR DESCRIPTION
**Pull request by woojiq**
The keyd package already exists but without a systemd service.

###### Description of changes
Add systemd service for the keyd package, and write a test.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [x] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

Also, I tested manually.

### CODE REVIEW

**Comment by ybnesm** [comment]
I added basic hardening to the systemd service. Now `systemd-analyze security keyd.service` looks like this:
```
  NAME                                                        DESCRIPTION                                                        EXPOSURE
✗ RootDirectory=/RootImage=                                   Service runs within the host's root directory                           0.1
  SupplementaryGroups=                                        Service runs as root, option does not matter
  RemoveIPC=                                                  Service runs as root, option does not apply
✗ User=/DynamicUser=                                          Service runs as root user                                               0.4
✓ RestrictRealtime=                                           Service realtime scheduling access is restricted
✓ CapabilityBoundingSet=~CAP_SYS_TIME                         Service processes cannot change the system clock
✓ NoNewPrivileges=                                            Service processes cannot acquire new privileges
✓ AmbientCapabilities=                                        Service process does not receive ambient capabilities
✗ PrivateDevices=                                             Service potentially has access to hardware devices                      0.2
✗ SystemCallArchitectures=                                    Service may execute system calls with all ABIs                          0.2
✗ RestrictAddressFamilies=~AF_PACKET                          Service may allocate packet sockets                                     0.2
✗ RestrictAddressFamilies=~AF_NETLINK                         Service may allocate netlink sockets                                    0.1
✗ RestrictAddressFamilies=~AF_UNIX                            Service may allocate local sockets                                      0.1
✗ RestrictAddressFamilies=~…                                  Service may allocate exotic sockets                                     0.3
✗ RestrictAddressFamilies=~AF_(INET|INET6)                    Service may allocate Internet sockets                                   0.3
✓ ProtectSystem=                                              Service has strict read-only access to the OS file hierarchy
✓ CapabilityBoundingSet=~CAP_SYS_RAWIO                        Service has no raw I/O access
✓ CapabilityBoundingSet=~CAP_SYS_PTRACE                       Service has no ptrace() debugging abilities
✓ CapabilityBoundingSet=~CAP_SYS_(NICE|RESOURCE)              Service has no privileges to change resource use parameters
✓ CapabilityBoundingSet=~CAP_NET_ADMIN                        Service has no network configuration privileges
✓ CapabilityBoundingSet=~CAP_NET_(BIND_SERVICE|BROADCAST|RAW) Service has no elevated networking privileges
✗ DeviceAllow=                                                Service has no device ACL                                               0.2
✓ CapabilityBoundingSet=~CAP_AUDIT_*                          Service has no audit subsystem access
✓ CapabilityBoundingSet=~CAP_SYS_ADMIN                        Service has no administrator privileges
✓ PrivateNetwork=                                             Service has no access to the host's network
✓ PrivateTmp=                                                 Service has no access to other software's temporary files
✓ CapabilityBoundingSet=~CAP_SYSLOG                           Service has no access to kernel logging
✓ ProtectHome=                                                Service has no access to home directories
✗ ProcSubset=                                                 Service has full access to non-process /proc files (/proc subset=)      0.1
✓ KeyringMode=                                                Service doesn't share key material with other services
✓ Delegate=                                                   Service does not maintain its own delegated control group subtree
✓ PrivateUsers=                                               Service does not have access to other users
✗ SystemCallFilter=~@clock                                    Service does not filter system calls                                    0.2
✗ SystemCallFilter=~@cpu-emulation                            Service does not filter system calls                                    0.1
✗ SystemCallFilter=~@debug                                    Service does not filter system calls                                    0.2
✗ SystemCallFilter=~@module                                   Service does not filter system calls                                    0.2
✗ SystemCallFilter=~@mount                                    Service does not filter system calls                                    0.2
✗ SystemCallFilter=~@obsolete                                 Service does not filter system calls                                    0.1
✗ SystemCallFilter=~@privileged                               Service does not filter system calls                                    0.2
✗ SystemCallFilter=~@raw-io                                   Service does not filter system calls                                    0.2
✗ SystemCallFilter=~@reboot                                   Service does not filter system calls                                    0.2
✗ SystemCallFilter=~@resources                                Service does not filter system calls                                    0.2
...skipping...
✗ SystemCallFilter=~@raw-io                                   Service does not filter system calls                                    0.2
✗ SystemCallFilter=~@reboot                                   Service does not filter system calls                                    0.2
✗ SystemCallFilter=~@resources                                Service does not filter system calls                                    0.2
✗ SystemCallFilter=~@swap                                     Service does not filter system calls                                    0.2
✗ IPAddressDeny=                                              Service does not define an IP address allow list                        0.2
✓ NotifyAccess=                                               Service child processes cannot alter service state
✓ ProtectClock=                                               Service cannot write to the hardware clock or system clock
✓ CapabilityBoundingSet=~CAP_KILL                             Service cannot send UNIX signals to arbitrary processes
✓ ProtectKernelLogs=                                          Service cannot read from or write to the kernel log ring buffer
✓ CapabilityBoundingSet=~CAP_(DAC_*|FOWNER|IPC_OWNER)         Service cannot override UNIX file/IPC permission checks
✓ ProtectControlGroups=                                       Service cannot modify the control group file system
✓ CapabilityBoundingSet=~CAP_LINUX_IMMUTABLE                  Service cannot mark files immutable
✓ CapabilityBoundingSet=~CAP_IPC_LOCK                         Service cannot lock memory into RAM
✓ ProtectKernelModules=                                       Service cannot load or read kernel modules
✓ CapabilityBoundingSet=~CAP_SYS_MODULE                       Service cannot load kernel modules
✓ CapabilityBoundingSet=~CAP_SYS_BOOT                         Service cannot issue reboot()
✓ CapabilityBoundingSet=~CAP_SYS_CHROOT                       Service cannot issue chroot()
✓ PrivateMounts=                                              Service cannot install system mounts
✓ MemoryDenyWriteExecute=                                     Service cannot create writable executable memory mappings
✓ RestrictNamespaces=~user                                    Service cannot create user namespaces
✓ RestrictNamespaces=~pid                                     Service cannot create process namespaces
✓ RestrictNamespaces=~net                                     Service cannot create network namespaces
✓ RestrictNamespaces=~uts                                     Service cannot create hostname namespaces
✓ RestrictNamespaces=~mnt                                     Service cannot create file system namespaces
✓ CapabilityBoundingSet=~CAP_MKNOD                            Service cannot create device nodes
✓ RestrictNamespaces=~cgroup                                  Service cannot create cgroup namespaces
✓ RestrictNamespaces=~ipc                                     Service cannot create IPC namespaces
✓ ProtectHostname=                                            Service cannot change system host/domainname
✓ CapabilityBoundingSet=~CAP_(CHOWN|FSETID|SETFCAP)           Service cannot change file ownership/access mode/capabilities
✓ CapabilityBoundingSet=~CAP_SET(UID|GID|PCAP)                Service cannot change UID/GID identities/capabilities
✓ LockPersonality=                                            Service cannot change ABI personality
✓ ProtectKernelTunables=                                      Service cannot alter kernel tunables (/proc/sys, …)
✓ CapabilityBoundingSet=~CAP_MAC_*                            Service cannot adjust SMACK MAC
✓ RestrictSUIDSGID=                                           SUID/SGID file creation by service is restricted
✗ ProtectProc=                                                                                                                        0.1
✓ CapabilityBoundingSet=~CAP_SYS_PACCT                        Service cannot use acct()
✓ CapabilityBoundingSet=~CAP_WAKE_ALARM                       Service cannot program timers that wake up the system
✓ CapabilityBoundingSet=~CAP_SYS_TTY_CONFIG                   Service cannot issue vhangup()
✓ CapabilityBoundingSet=~CAP_BLOCK_SUSPEND                    Service cannot establish wake locks
✓ CapabilityBoundingSet=~CAP_LEASE                            Service cannot create file leases
✗ UMask=                                                      Files created by service are world-readable by default                  0.1
→ Overall exposure level for keyd.service: 3.5 OK 🙂
```
However, I'm not a a systemd expert, so there are options for further improvements.

Some observation:
* It must be a root user (`User=/DynamicUser=`)
* It must have access to hardware devices (`PrivateDevices`)

**Comment by ybnesm** [comment]
cc @pennae could you review the PR, please? You have already watched half on the matrix server)

File: nixos/modules/services/hardware/keyd.nix [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+      restartTriggers = [
+        config.environment.etc."keyd/default.conf".source
+      ];
+
+      serviceConfig = {
+        ExecStart = "${pkgs.keyd}/bin/keyd";
+        Restart = "always";
+
+        # Hardening
+        CapabilityBoundingSet = "";
```
[the remark this answers is not in the document]
Comment (ybnesm) [reply nixos/modules/services/hardware/keyd.nix:68]: > keyd attempts to solve this problem by providing a flexible **_system wide_** daemon which remaps keys using kernel level input primitives (evdev, input).

Afaik, system-wide services can only run as root. Did you mean this?

**Comment by ybnesm** [comment]
I replaced `extraConfig` with next options: `settings` (ini format) and `ids` (list of string). Is [this solution](https://github.com/NixOS/nixpkgs/pull/221321/files#diff-24d39c72c95a447767bc244d33dd85eafef2954906458fcb66ae4b85092a10d0R6-R10) good? And corrected other recommendations, except hardening.

> can this service run without root, but with some capabilities or supplementary groups to grant it access to the necessary devices?

I didn't quite understand what was meant. `Keyd` is a system-wide service, so it cannot be run as `systemd --user...`.

P.S. Thanks for the detailed review.

**Comment by ybnesm** [comment]
> systemd _system_ services can be run as arbitrary users/groups by setting `User/Group/DynamicUser`. this is different from systemd _user_ services which run in the context of a user service manager and can only run as that user (to first approximation). `keyd` should continue to run as a system service, but not with uid/gid 0 if that's at all possible. (that kanata can do it is a sign in favor of this)

I am not sure if this is possible due to the following reasons which I have researched. I wrote this as in kanata (only this, without extra hardening):
``` nix
hardware.uinput.enable = true;
...
{
DynamicUser = true;
DeviceAllow = [
    "/dev/uinput rw"
    "char-input r"
];
SupplementaryGroups = with config.users.groups; [
    input.name
    uinput.name
];
```
But it didn't work because of the error: `failed to open /dev/input/event-*. Permission denied`. So I added `"/dev/input rw"` to the allowed devices. The error remained. I tried writing 20 entries like this `/dev/input/event0 /dev/input/event1 ...`. Almost all errors are gone, but two errors remain:
```
failed to open /dev/input/event11
failed to open /dev/input/event10
```
I looked at `ls -l /dev/input/`, these events are gone, however, they are present without the `DynamicUser` option. Is this a sign that `dynamicuser` cannot be applied here? I've read a couple of articles about DynamicUser options, reviewed similar files inside the `nixpkgs` repository, and it looks like this could be the answer, but it's probably not an "error" on the systemd side.

If necessary, I can try to ask the `keyd` developer about it.

**Comment by ybnesm** [comment]
I tested with minimum strengthening parameters. First I tried this:
``` nix
serviceConfig = {
  ExecStart = "${pkgs.keyd}/bin/keyd";
  Restart = "always";

  DynamicUser = true;
  DeviceAllow = [ "/dev/uinput rw" ];
  SupplementaryGroups = with config.users.groups; [
    input.name
  ];
};
```
`uinput permission denied` -> add `uinput` to supplementarygroups and `hardware.uinput.enable = true;`
`faile to open /dev/input/event*. Permission denied` -> add `DeviceAllow = ["/dev/input rw"];` 
The error remains. I looked on the internet (and chatgpt:) how I can fix this, there is a `chmod` solution, but it looks unsound. I even tried add my user to input/uinput group, but it didn't help.

**Comment by ybnesm** [comment]
I give up. I don't have enough time/patience anymore to complete this task. I asked similar question here https://github.com/rvaiya/keyd/issues/459. From what I read: 
> but it should still be run as root since it needs access to your input devices and /dev/uinput. The best way to achieve this is to have your init system start and manage it at boot time.

Although `SupplementaryGroups` and `DeviceAllow` should have helped, they didn't :cry: 

If you accept the current behavior I can squash commits before merging, if not - feel free to close the PR I won't be angry. Anyway, I gained vast knowledge doing this stuff 😌

**Comment by ybnesm** [comment]
A few last questions 🧐.
1. [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md)
> Pull requests should not be squash merged in order to keep complete commit messages and GPG signatures intact and must not be when the change doesn't make sense as a single commit.

Should I squash all the commits into one and keep the commit messages, or should I just squash my commits and keep yours separately?

2. Now service is working well, but it feels a little bit of magic for me, especially these two things: `KEYD_SOCKET = "/run/keyd/keyd.sock"` and `"char-input rw"`. Where did you find information about them? Recursive search inside the keyd repo hasn't found any mentions for `KEYD_SOCKET` only `SOCKET_PATH` inside Makefile. Also, `char-input rw` sounds intuitive but in google I can't find any related information. There is about [`char-rtc`](https://www.freedesktop.org/software/systemd/man/systemd.exec.html#ProtectClock=) but there is no about `char-input`.

**Comment by ybnesm** [comment]
I think this PR is now ready for landing. If I misunderstood what you said, I can roll back the commit history using a local copy or the `git reflog` command)

**Review by pennae — APPROVED** [review]

**Pull request merged**